munge/srseng/dictionary.ts

```diff
@@ -0,0 +1,245 @@
+import assert from "assert";
+import { minBy, sortBy } from "lodash";
+import { DefaultMap, makePrettierJson, stringDistance } from "./util";
+
+export type Analysis = [string[], string, string[]];
+type DefinitionList = {
+  definition: string;
+  sources: string[];
+}[];
+
+class DictionaryEntry {
+  head?: string;
+  analysis?: Analysis;
+  paradigm?: string;
+  senses?: DefinitionList;
+  slug?: string;
+
+  addDefinition(definition: string) {
+    if (!definition.trim()) {
+      return;
+    }
+
+    if (this.senses === undefined) {
+      this.senses = [];
+    }
+    for (const k of this.senses) {
+      if (k.definition === definition) {
+        return;
+      }
+    }
+    this.senses.push({ definition, sources: ["OS"] });
+  }
+
+  fstLemma() {
+    assert(this.analysis);
+    return this.analysis[1];
+  }
+}
+
+class Wordform {
+  head?: string;
+  analysis?: Analysis;
+  senses?: DefinitionList;
+  formOf?: DictionaryEntry;
+}
+
+type ExportableWordform = Required<Omit<Wordform, "formOf">> & {
+  formOf: string;
+};
+
+export class Dictionary {
+  /**
+   * FST tags which distinguish the lexeme, e.g., +N and +V, as opposed to tags
+   * that distinguish the wordform within a lexeme, e.g., +Sg and +Pl.
+   */
+  readonly lexicalTags: Set<string>;
+  _entries: (DictionaryEntry | Wordform)[];
+  _byText: Map<string, DictionaryEntry>;
```

Comment:
[Private class fields](https://github.com/tc39/proposal-class-fields) are stable in Node [as of v14](https://node.green/#ES2022-features-instance-class-fields), if that tickles your fancy.

```suggestion
  #entries = (DictionaryEntry | Wordform)[];
  #byText = Map<string, DictionaryEntry>;
```

Reply:
Thanks for the suggestion. I’m not sold on private class fields; I can’t use them as a general replacement for `_` variables because private class fields can’t be accessed from subclasses, and sometimes I create subclasses that need to access internal variables of the superclass(es).

TypeScript has `private` and `protected` keywords that I could use if I wanted more protection here. Those keywords aren’t enforced at runtime, and there are workarounds, but TypeScript will ddefinitely complain if you accidentally violate them. Even those keywords normally don’t seem worth the effort for what I’m usually doing.